databuilder/databuilder/models/type_metadata.py

```diff
@@ -0,0 +1,386 @@
+            assert description_key is not None, f"Could not retrieve description key for {self.name}"
+            yield self.description.get_relation(
+                TypeMetadata.NODE_LABEL,
+                self.key(),
+                description_key
+            )
+
+        if not self.is_terminal_type():
+            assert self.array_inner_type is not None, f"Array inner type must be set for {self.name}"
+            yield from self.array_inner_type.create_relation_iterator()
+
+
+class MapTypeMetadata(TypeMetadata):
+    kind = 'map'
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super(MapTypeMetadata, self).__init__(*args, **kwargs)
+        self.map_key_type: Optional[TypeMetadata] = None
+        self.map_value_type: Optional[TypeMetadata] = None
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, MapTypeMetadata):
+            return (self.name == other.name and
+                    self.map_key_type == other.map_key_type and
+                    self.map_value_type == other.map_value_type and
+                    self.type_str == other.type_str and
+                    self.description == other.description and
+                    self.sort_order == other.sort_order and
+                    self.key() == other.key())
+        return False
+
+    def is_terminal_type(self) -> bool:
+        return not self.map_key_type or not self.map_value_type
+
+    def create_node_iterator(self) -> Iterator[GraphNode]:
+        assert self.map_key_type is not None, f"Map key type must be set for {self.name}"
+        assert self.map_value_type is not None, f"Map value type must be set for {self.name}"
+
+        node_attributes: Dict[str, Union[str, None, int]] = {
+            TypeMetadata.KIND: self.kind,
+            TypeMetadata.NAME: self.name,
+            TypeMetadata.DATA_TYPE: self.type_str,
+            TypeMetadata.MAP_KEY: self.map_key_type.type_str,
+            TypeMetadata.MAP_VALUE: self.map_value_type.type_str
+        }
+
+        if isinstance(self.sort_order, int):
+            node_attributes[TypeMetadata.SORT_ORDER] = self.sort_order
+
+        yield GraphNode(
+            key=self.key(),
+            label=TypeMetadata.NODE_LABEL,
+            attributes=node_attributes
+        )
+
+        if self.description:
+            description_key = self.description_key()
+            assert description_key is not None, f"Could not retrieve description key for {self.name}"
+            yield self.description.get_node(description_key)
+
+        if not self.is_terminal_type():
+            yield from self.map_key_type.create_node_iterator()
+            yield from self.map_value_type.create_node_iterator()
+
+    def create_relation_iterator(self) -> Iterator[GraphRelationship]:
+        assert self.map_key_type is not None, f"Map key type must be set for {self.name}"
+        assert self.map_value_type is not None, f"Map value type must be set for {self.name}"
```

Comment:
this seems wrong (likewise for `create_node_iterator`). If the node is terminal and the key/values are not set, the node still needs to be GraphSerializable. I.e. `create_node_iterator` would just create it's own node, and `create_relation_iterator` would just create the relationship between parent and this node, and to description node.